### PR TITLE
Change CacheState.register to a fluent API

### PR DESCRIPTION
**Pull request by pkoenig10**
Follow up from https://github.com/palantir/tritium/pull/1919

This is a slightly less obvious but more ergonomic API for consumers.

### CODE REVIEW

File: tritium-caffeine/src/main/java/com/palantir/tritium/metrics/caffeine/CacheStats.java [hunk 1]
```diff
@@ -68,12 +68,24 @@ public static CacheStats of(TaggedMetricRegistry taggedMetricRegistry, @Safe Str
...
+     *     Cache&lt;Integer, String&gt; cache = CacheStats.of(taggedMetricRegistry, "your-cache-name")
+     *             .register(stats -> Caffeine.newBuilder()
+     *                     .recordStats(stats)
+     *                     .build());
+     * </pre>
+     * @param cacheFactory method which will be invoked to construct the cache
+     * @return the constructed cache instance
      */
-    public <K, V, C extends Cache<K, V>> C register(C cache) {
+    public <K, V, C extends Cache<K, V>> C register(Function<CacheStats, C> cacheFactory) {
```
Comment (carterkozak) [inline tritium-caffeine/src/main/java/com/palantir/tritium/metrics/caffeine/CacheStats.java:86]: nice

**Review by carterkozak — APPROVED** [review]

**Comment by pkoenig10** [comment]
One nice thing about this API is that it makes it difficult to forget to call `recordStats`, since the provided `CacheStats` variable will be unused otherwise.

**Comment by carterkozak** [comment]
Before this rolls out broadly, could we add a check to throw if stats aren't being recorded (which implies the cacheStats value hasn't been provided to the builder)?

**Comment by pkoenig10** [comment]
https://github.com/palantir/tritium/pull/1922

**Pull request merged**